### ISSUE TITLE
List of all keywords in crystal lang

### ISSUE THREAD

**Issue by scriptnull** [crystal-lang/crystal-book#124]
It would be helpful to have all the keywords of crystal language in a page. 

Examples:
- https://docs.oracle.com/javase/tutorial/java/nutsandbolts/_keywords.html
- https://golang.org/ref/spec#Keywords

It would be helpful for people coming from other ecosystems to get the idea of available keywords.

**Comment by faustinoaq** [comment]
Hi @scriptnull see: https://github.com/crystal-lang/crystal/wiki/Crystal-for-Rubyists#available-keywords

**Comment by RX14** [comment]
Looks out of date to me, `union` isn't a keyword.

**Comment by faustinoaq** [comment]
> Looks out of date to me, union isn't a keyword.

I fixed it 👍

**Comment by scriptnull** [comment]
Hi guys,

Is there a place in the source code (https://github.com/crystal-lang/crystal) that could help me collect all the latest keywords in the language? 

I did little research on this and the only place I could arrive at is 

https://github.com/crystal-lang/crystal/blob/ebe683c5eb9b85da55859e8c1b1fecfbf9bb3da2/spec/compiler/lexer/lexer_spec.cr#L143-L148

I could also see the presence of union keyword in the spec. ( is it left out as a mistake or intentionally ? )

Any pointers on this would help me create the documentation page and serve as a good starting point for reading the crystal source code. 

Thanks.

**Comment by faustinoaq** [comment]
@scriptnull I updated the wiki https://github.com/crystal-lang/crystal/wiki/Crystal-for-Rubyists#available-keywords

Thanks you for the link to `lexer_spec.cr` 👍

**Comment by bew** [comment]
`union` is a keyword for writing lib bindings (doc: https://crystal-lang.org/docs/syntax_and_semantics/c_bindings/union.html)

**Comment by RX14** [comment]
@bew oh yeah, i forgot about that.

This issue is that there is no magical list of keywords? Keywords in which context? In method bodies, `if = true` doesn't compile, but `union = true` [does](https://carc.in/#/r/2vls). However, in `lib` context, `union` is clearly a keyword. In class context, `union` is [not a keyword](https://carc.in/#/r/2vlx). For example, all of these keywords can be [used as method names](https://carc.in/#/r/2vly) *if* the method call has a reciever. What is and is not a keyword in crystal highly depends on the context. There is no answer.

**Comment by scriptnull** [comment]
@RX14 Wow! Thanks for sharing the information. 

I am quite surprised with the fact that crystal's keywords differ based on the context in which they are used. I could get the advantages of it by your `union` example. 

But, from all the programming languages I have worked with, every single one of them uses a fixed and standard list of keywords, which could not be used in places other than their designated places ( Example: Cannot define `for` as function name in Go, even if it has a receiver )

If you are aware of how this language design choice was made, please point me to the RFC/design document for it. I am interested in learning more about it.

Furthermore, the thing that I am afraid is, with more keywords coming in and going away in future releases of crystal, the language users might get confusions regarding the usage of various keywords in their contexts and out of their contexts. So it would be really helpful, if someone who has idea about how keywords in crystal work writes a brief summary about it.

Please note that the origin of this issue is me ( as a beginner ) struggling for about an hour to use `next` as a property in my class. 

```crystal
# https://carc.in/#/r/2vtt
# Does not compile
class Foo
  @next : Foo | Nil
  
  property next
end
```

```crystal
# https://carc.in/#/r/2vtu
# compiles correctly
class Foo
  @next : Foo | Nil
  
  property "next"
end
```

I am not from ruby background and I didn't even think that `next` could be a keyword. ( more used to `continue` 😄  ). And when I searched for keywords in crystal, I couldn't get to it ( because the 'magical' page does not exist ), but thankfully there is a reference about `next` keyword in crystal book.

Anyways, it's ok to not have an entire list of keywords, but it would be nicer to have something that explains this context based keyword behavior in crystal somewhere ( which might actually help beginners like me to get a nice idea about it and spend less time making mistakes )

**Comment by RX14** [comment]
I think the most useful list of keywords would be the list of all possible keywords in any context. 

I don't know the exact reason why keywords depend on context, perhaps @asterite can tell us if it just fell out of using a custom written lexer/parser or was a conscious decision.

**Comment by Metloff** [comment]
Can we make file with keywords? For example in Ruby we have:
https://github.com/ruby/ruby/blob/trunk/defs/keywords

This will help solve problems like:
https://github.com/amberframework/amber/issues/995

**Comment by HertzDevil** [comment]
The current list of identifiers that could possibly be keywords according to the lexer are:

* `abstract`
* `alias`
* `annotation`
* `as`
* `as?`
* `asm`
* `begin`
* `break`
* `case`
* `class`
* `def`
* `do`
* `else`
* `elsif`
* `end`
* `ensure`
* `enum`
* `extend`
* `false`
* `for`
* `fun`
* `if`
* `in`
* `include`
* `instance_sizeof`
* `is_a?`
* `lib`
* `macro`
* `module`
* `next`
* `nil`
* `nil?`
* `of`
* `offsetof`
* `out`
* `pointerof`
* `private`
* `protected`
* `require`
* `rescue`
* `responds_to?`
* `return`
* `select`
* `self`
* `sizeof`
* `struct`
* `super`
* `then`
* `true`
* `type`
* `typeof`
* `uninitialized`
* `union`
* `unless`
* `until`
* `verbatim`
* `when`
* `while`
* `with`
* `yield`

A notable exception is `forall`.

**Comment by zw963** [comment]
> 



> The current list of identifiers that could possibly be keywords according to the lexer are:

I create a PR for emacs company-mode for add those keywords for emacs crystal-mode.

https://github.com/company-mode/company-mode/pull/1338

Can i know why `forall` is not consider as a keyword?

**Comment by nobodywasishere** [comment]
> Can i know why forall is not consider as a keyword?

It's not a [keyword type](https://github.com/crystal-lang/crystal/blob/cc30da2d3b8d5d3ee300a65d73268b0ded5638f3/src/compiler/crystal/syntax/token.cr#L5) for the lexer/parser, and is instead [checked explicitly](https://github.com/crystal-lang/crystal/blob/cc30da2d3b8d5d3ee300a65d73268b0ded5638f3/src/compiler/crystal/syntax/parser.cr#L3709) when parsing defs

**Comment by straight-shoota** [comment]
Yeah, I think this is merely a technicality. For the end result it makes no difference if the compiler parses `forall` as a special token or not.
